src/QueryHandler.cc

```diff
@@ -368,13 +370,32 @@ void QueryHandler::process_query(protobufs::queryMessage& proto_query,
 
         Json::Value& tx_responses = pmgd_query.run();
 
-        if (tx_responses.size() != root.size()) { // error
+        if (!tx_responses.isArray()) {
+            Json::StyledWriter writer;
+            std::cerr << "PMGD Response:" << std::endl;
+            std::cerr << writer.write(tx_responses) << std::endl;
+
+            std::string tx_error_msg = "Failed PMGD Transaction";
+            if (tx_responses.isMember("info")) {
+                tx_error_msg += ": " + tx_responses["info"].asString();
+            }
+
+            cmd_result["status"] = RSCommand::Error;
+            cmd_result["info"] = tx_error_msg;
+
             cmd_current = "Transaction";
-            cmd_result = tx_responses;
-            cmd_result["info"] = "Failed PMGDTransaction";
+            error(cmd_result, cmd_current);
+            return;
+        }
+        else if (tx_responses.size() != root.size()) {
+            Json::StyledWriter writer;
+            std::cerr << "PMGD Response:" << std::endl;
+            std::cerr << writer.write(tx_responses) << std::endl;
+
             cmd_result["status"] = RSCommand::Error;
-            Json::StyledWriter w;
-            std::cerr << w.write(tx_responses);
+            cmd_result["info"] = "Failed PMGD Transaction: Wrong Response Size";
+
+            cmd_current = "Transaction";
```

Comment:
There seems to be a lot of commonality between the two error flows. Could they be combined, using
`if (!tx_responses.isArray() || tx_responses.size() != root.size())`